contracts/Masks/DEFYMasks.sol

```diff
@@ -0,0 +1,186 @@
+// | | | |  __||  _|   \ /
+// | |/ /| |___| |     | |
+// |___/ \____/\_|     \_/
+//
+// WELCOME TO THE REVOLUTION
+
+contract DEFYMasks is
+    ERC721,
+    ERC721Enumerable,
+    Pausable,
+    Ownable,
+    AccessControl,
+    DefaultOperatorFilterer
+{
+    using Counters for Counters.Counter;
+
+    bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+    bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+    bytes32 public constant BURNER_ROLE = keccak256("BURNER_ROLE");
+
+    // Permits an address to transfer below 1 mask
+    mapping(address => bool) private _tokenTradingEnabled;
+
+    Counters.Counter private _tokenIdCounter;
+
+    // Base URI for mask token uris
+    string private _maskBaseURI;
+
+    /// @notice Sets the base URI used for the tokens.
+    function setBaseURI(string memory uri) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        _maskBaseURI = uri;
+    }
+
+    /// @notice Get the TokenURI for the supplied token, in the form {baseURI}{tokenId}
+    function tokenURI(uint256 tokenId)
+        public
+        view
+        override
+        returns (string memory)
+    {
+        require(_exists(tokenId), "DEFYMasks: URI query for nonexistent token");
+
+        return
+            string(
+                abi.encodePacked(
+                    _maskBaseURI,
+                    Strings.toString(tokenId),
+                    ".json"
+                )
+            );
+    }
+
+    constructor() ERC721("DEFYMasks", "DM") {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(PAUSER_ROLE, msg.sender);
+    }
+
+    function safeMint(address to) public onlyRole(MINTER_ROLE) {
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(to, tokenId);
+    }
+
+    function safeBatchMint(address[] calldata to) public onlyRole(MINTER_ROLE) {
+        for (uint256 i = 0; i < to.length; i++) {
+            safeMint(to[i]);
+        }
+    }
+
+    function burnMask(uint256 tokenId) public onlyRole(BURNER_ROLE) {
+        _burn(tokenId);
+    }
+
+    function batchBurnMasks(uint256[] calldata tokenIds)
+        public
+        onlyRole(BURNER_ROLE)
+    {
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            _burn(tokenIds[i]);
+        }
+    }
+
+    // Checks for token transfers and burns that either:
+    //  Admin enabled trading for operative address; or,
+    //  Operative Address will have at least 1 mask after transfer
+    // If check to allow minting to bypass check
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal override(ERC721, ERC721Enumerable) whenNotPaused {
+        if (tokenId < totalSupply()) {
+            require(
+                _tokenTradingEnabled[from] || balanceOf(from) >= 2,
+                "DEFYMasks: Token trading is not enabled"
+            );
+        }
+
+        super._beforeTokenTransfer(from, to, tokenId);
```

Comment:
Should disable tokenTradingEnabled for the address after the trade if the balance is now 0 after the trade, so that we don't have a situation where someone ends up with a special address that is permanently unlocked due to us forgetting to disable it again

Reply:
so run that in after token transfer?